### DISCUSSION TITLE
Table editor usability tweaks

### DISCUSSION THREAD

**Discussion by tridian-tn** [zim-desktop-wiki/zim-desktop-wiki#2564] [Ideas]
I was hoping to make some changes to the table editor, to make it easier to get around when a cell is being edited, without having to press return to get out of the editor. Specifically:

1. If the cursor is in a cell (in all but the last column), pressing tab moves the cursor to the next cell. 
2. If the cursor is in the last cell of a row but not on the last row, pressing tab moves the cursor to the first cell of the next row.
3. If the cursor is in the last cell of a row and it is the last row, pressing tab creates a new row and moves the cursor to the first cell of the new row
4. Pressing shift-tab moves the cursor to the previous cell in a similar way, but if the cursor is in the first cell (0, 0), it doesn't move.
5. Similarly, if the cursor is in a cell and given that Zim table cells don't support multiline text, pressing up or down moves the cursor in that direction. If the cursor was already on the first or last row....I'm not sure yet. Either the cursor stays where it is ("locked" in the table) or focus moves outside of the table editor to the core page editor.

I'm admittedly referencing OneNote's behaviour here (and Excel/LibreOffice Calc to a degree), but OneNote's table handling is the only thing about it I miss and I don't think that's purely based on familiarity: it works really well.

There are other things I can think of (e.g. off the top of my head, better colour differentiation between a selected row and the cell focus, or allowing a single-click to select another row when the cursor is already editing a cell) but I wanted to check first if there was any appetite for these kind of changes. 

I've got the tab cursor movement working to a basic level, so I'm in a good place to raise a draft PR and chip away at it.

Cheers

**Comment by introt** [comment]
There's indeed appetite for table editor usability tweaks, PRs are very welcome!

> Pressing shift-tab moves the cursor to the previous cell in a similar way, but if the cursor is in the first cell (0, 0), it doesn't move.

I guess moving focus between the table and the rest of the text buffer wouldn't work due to tab doubling as indent.

> Similarly, if the cursor is in a cell and given that Zim table cells don't support multiline text, pressing up or down moves the cursor in that direction.

There's a stale PR for multiline text: #1370 (+ #953). I'd go for enter and shift + enter instead (The Spreadsheet Way (TM)) - and even if we don't support multiline, I'd want to reserve up and down for wrapped lines (do they work atm? #1095).

> There are other things I can think of (e.g. off the top of my head, better colour differentiation between a selected row and the cell focus, or allowing a single-click to select another row when the cursor is already editing a cell) but I wanted to check first if there was any appetite for these kind of changes.

See "change table border color" #681, "[Table Editor] UI suggestion: double-click to immediately edit cell" #591 - for more inspiration, search "table" in issues, PRs and Discussions.

See also #2269 - I don't think the mentioned fix for #853 ever made it into the code base. Same with the tangentially table related "Uncaught exception when sorting "None" string in util.natural_sort_key" #1061 

Happy Hacking!

**Reply by tridian-tn** [reply]
Thanks! And thanks for the research too. Much appreciated. I'll gather together the various suggestions, issues and PRs and write a plan.
